creator-node/src/fileManager.js

```diff
@@ -95,57 +96,105 @@ async function saveFileToIPFSFromFS (req, srcPath, fileType, sourceFile, transac
-  req.logger.info(`checking if ${multihash} already available on local ipfs node`)
+  req.logger.debug(`checking if ${multihash} already available on local ipfs node`)
   try {
-    fileBuffer = await ipfs.cat(multihash)
-    req.logger.info(`Retrieved file for ${multihash} from local ipfs node`)
+    fileBuffer = await Utils.ipfsCat(multihash, req, 500)
+    req.logger.debug(`Retrieved file for ${multihash} from local ipfs node`)
+    // Write file to disk.
+    await writeFile(expectedStoragePath, fileBuffer)
+    req.logger.info(`wrote file to ${expectedStoragePath}, obtained via ipfs cat`)
   } catch (e) {
     req.logger.info(`Multihash ${multihash} is not available on local ipfs node`)
   }
 
   // If file not already available on local INode, fetch from IPFS.
   if (fileBuffer === null) {
-    req.logger.info(`Attempting to get ${multihash} from IPFS`)
+    req.logger.debug(`Attempting to get ${multihash} from IPFS`)
     let output
     try {
-      output = await ipfs.get(multihash)
+      // ipfsGet returns a BufferList object, not a buffer
+      // but still compatible into writeFile directly
+      output = await Utils.ipfsGet(multihash, req, 5000)
+      fileBuffer = output
+      req.logger.debug(`retrieved file for multihash ${multihash} from local ipfs node`)
+
+      // Write file to disk.      
+      const destinationStream = fs.createWriteStream(expectedStoragePath)
+      await new Promise((resolve, reject) => {
+        fileBuffer.pipe(destinationStream)
+        fileBuffer.on('end', () => { resolve() })
+        fileBuffer.on('error', err => { destinationStream.end(); reject(err) })
+      })
+      req.logger.info(`wrote file to ${expectedStoragePath}, obtained via ipfs get`)
     } catch (e) {
-      throw new Error(`Failed to retrieve file for multihash ${multihash} from IPFS`)
+      req.logger.info(`Failed to retrieve file for multihash ${multihash} from IPFS ${e.message}`)
     }
-    if (output.length !== 1) throw new Error('Audius track segment multihash must map to 1 file')
-    fileBuffer = output[0].content
-    req.logger.info(`retrieved file for multihash ${multihash} from path ${output[0].path}`)
   }
 
-  // Write file to disk.
-  const storagePath = path.join(req.app.get('storagePath'), multihash)
-  assert.strictEqual(storagePath, expectedStoragePath)
-  req.logger.info(`writing file to ${storagePath}...`)
-  await writeFile(storagePath, fileBuffer)
-  req.logger.info(`wrote file to ${storagePath}`)
+  // if file is still null, try to fetch from other cnode gateways with onlyFS=true
+  if (fileBuffer === null) {
+    try {
+      let response
+      // ..replace(/\/$/, "") removes trailing slashes
+      req.logger.debug(`Attempting to fetch multihash ${multihash} by racing replica set endpoints`)
+
+      // TODO - change this to proper directory gateway
+      const urls = gatewaysToTry.map(endpoint => `${endpoint.replace(/\/$/, '')}/ipfs/${multihash}`)
+
+      // TODO make this more parallel
```

Comment:
promise.race should work here right? we should try to get this in before merging

Reply:
so i attempted to use promise.race here initially but decided against it for two reasons. it's not natively supported in node yet and also we'd be making parallel requests to all replica set gateways. that's a level of load i don't know we should place on the gateways suddenly especially since this is an async operation in the background